query/marshal.go

```diff
@@ -0,0 +1,177 @@
+package query
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+)
+
+// We implement a custom binary marshaller for a list of repos to
+// branches. When profiling Sourcegraph this is one of the dominant items.
+//
+// Wire-format of map[string][]string is pretty straightforward:
+//
+// byte(1) version
+// uvarint(len(map))
+// for k, vs in map:
+//   str(k)
+//   uvarint(len(vs))
+//   for v in vs:
+//     str(v)
+//
+//  where str(v) is uvarint(len(v)) bytes(v)
+//
+// The above format gives about the same size encoding as gob does. However,
+// gob doesn't have a specialization for map[string][]string so we get to
+// avoid a lot of intermediate allocations.
+//
+// The only other specialization we add is treating []string{"HEAD"} as if it
+// was []string{}. This is the most common value for branches so avoids the
+// need to write it on the wire. This makes us beat gob for encoded size.
+//
+// The above adds up to a huge improvement, worth the extra complexity:
+//
+// name                   old time/op    new time/op    delta
+// RepoBranches_Encode-8    2.21ms ± 3%    0.63ms ± 5%  -71.55%  (p=0.000 n=10+10)
+// RepoBranches_Decode-8    4.08ms ± 5%    1.20ms ± 1%  -70.58%  (p=0.000 n=10+10)
+//
+// name                   old bytes      new bytes      delta
+// RepoBranches_Encode-8     393kB ± 0%     344kB ± 0%  -12.48%  (p=0.000 n=10+10)
+//
+// name                   old alloc/op   new alloc/op   delta
+// RepoBranches_Encode-8     726kB ± 0%     344kB ± 0%  -52.59%  (p=0.000 n=6+9)
+// RepoBranches_Decode-8    2.31MB ± 0%    1.58MB ± 0%  -31.62%  (p=0.000 n=10+10)
+//
+// name                   old allocs/op  new allocs/op  delta
+// RepoBranches_Encode-8     20.0k ± 0%      0.0k ± 0%  -99.99%  (p=0.000 n=10+10)
+// RepoBranches_Decode-8     50.6k ± 0%     20.8k ± 0%  -58.94%  (p=0.000 n=10+10)
+
+// repoBranchesEncode implements an efficient encoder for RepoBranches.
+func repoBranchesEncode(repoBranches map[string][]string) ([]byte, error) {
+	var b bytes.Buffer
+	var enc [8]byte
+	varint := func(n int) {
+		m := binary.PutUvarint(enc[:], uint64(n))
+		b.Write(enc[:m])
+	}
+	str := func(s string) {
+		varint(len(s))
+		b.WriteString(s)
+	}
+	strSize := func(s string) int {
+		return binary.PutUvarint(enc[:], uint64(len(s))) + len(s)
+	}
+
+	// Version
+	b.WriteByte(1)
+
+	// Length
+	varint(len(repoBranches))
+
+	// Calculate size
+	size := 0
+	for name, branches := range repoBranches {
+		size += strSize(name) + 1
+		if l := len(branches); l == 1 && branches[0] == "HEAD" {
+			continue
+		} else if l > 255 {
```

Comment:
Where does the limit l>255 come from? The encoding per se doesn't seem to require any limit, does it?

Reply:
Later on I encode l into a byte. I could just check that it is less than 64 like I mention below. I'll make the comments clearer.

FYI zoekt only supports indexing upto 64 branches (it uses a bitmask on a 64-bit integer to encode branch info per document)